examples/ecs/advanced_state.rs

```diff
@@ -0,0 +1,347 @@
+//! This example illustrates how to use some of the more advance [`States`] functionality for high-level app control flow.
+//!
+//! The use case here is identical to the regular [state](./state.rs) example, but we will be
+//! utilizing the `StateMatcher` trait more directly.
+use bevy::prelude::*;
+
+fn main() {
+    App::new()
+        .add_plugins(DefaultPlugins)
+        .add_state::<AppState>()
+        .add_systems(Startup, setup)
+        // When we check whether a state is valid, we use a `StateMatcher`.
+        // In the simple example, we use pattern matching to generate one on the fly
+        // But we can also use pre-defined ones, like `ShowsUI`
+        .add_systems(on_exit_strict!(ShowsUI), cleanup_ui)
+        // In addition, all `States` also implement `StateMatcher`, so you can use them here.
+        // Although there are some optimizations around `OnEnter(S)` and `OnExit(S)` that make
+        // those a better choice in this case.
+        .add_systems(on_enter!(AppState::Menu), setup_menu)
+        // A state matcher can even be an enum! or really anything with that implements
+        // the state matcher trait (see below).
+        .add_systems(on_enter!(InGame::Paused), setup_paused)
+        .add_systems(on_enter!(InGame::Running), setup_in_game_ui)
+        // And unlike a regular state, the values are not mutually exclusive.
+        // So this will be valid both when the game is paused and when it's not
+        .add_systems(on_enter!(InGame::Any), setup_game)
+        // And all the same things apply with the `in_state!` macro
+        .add_systems(Update, menu.run_if(in_state!(AppState::Menu)))
+        .add_systems(Update, change_color.run_if(in_state!(InGame::Any)))
+        .add_systems(Update, toggle_pause)
+        // And it still works with conditional states as well
+        .add_sub_state::<MovementState, _>(InGame::Running)
+        .add_systems(Update, invert_movement.run_if(in_state!(MovementState, _)))
+        .add_systems(Update, movement.run_if(in_state(MovementState::Normal)))
+        .add_systems(
+            Update,
+            inverted_movement.run_if(in_state(MovementState::Invert)),
+        )
+        .run();
+}
+
+// The simplest way to define a state matcher is using the `state_matcher!` macro.
+// You pass in the visibility (optionally), name, state type and match expression.
+state_matcher!(pub ShowsUI, AppState, Menu | InGame { .. } );
```

Comment:
It isn't clear what exactly this macro is doing.

Reply:
I was going back and forth on how do define this macro... It's main purpose is to create a struct with an implementation of StateMatcher based on the provided pattern.

Originally I was thinking of using a Derive, but I worried that it would require having the matching occur in an attribute, which just felt... worse?

Just for clarity, I'll try to illustrate what this creates, and some of the options I thought of:

```rust
#[derive(Debug, Eq, PartialEq, Hash, Clone)]
pub struct ShowsUI;

impl StateMatcher<AppState> for ShowsUI {
  fn match_state(&self, state: &AppState) -> bool {
    match state {
      AppState::Menu | AppState::InGame { .. } => true,
      _ => false
    }
  }
}
```

The main other contender I thought of:
```rust
#[derive(Debug, Eq, PartialEq, Hash, Clone, StateMatcher)]
#[match(AppState, Menu | InGame { .. })]
pub struct ShowsUI;
```

And another version that would have a more complex derive, and would be able to build on the previous one but provide some added options, using the logic around `InGame` below:
```rust
#[derive(Debug, Eq, PartialEq, Hash, Clone, StateMatcher)]
#[match(AppState)]
pub enum  InGame{
  #[match(AppState::InGame { .. })]
  Any,
  #[match(AppState::InGame { paused: false } )]
  Running,
  #[match(AppState::InGame { paused: true } )]
  Paused
} 
```

Reply:
Since it's now easier to actually use raw function pointers for state matching, I don't think this is needed anymore.

Instead, you could do:

```rust
fn shows_ui(state: &AppState) -> bool {
  matches!(state, AppState::Menu | AppState::InGame { .. })
}
```

this is more verbose, but it is also easier to tell what's happening, and where the flexibility lies. I will remove the macro, but can bring it back if asked at a later date.

Reply:
Turns out, my attempted solution here won't work - `fn` requires explicit coercion, which actually makes using this really annoying... either:
```rust 
on_exit_strict!(shows_ui as fn(&AppState) -> bool)
```
or
```rust
const SHOWS_UI: fn(&AppState) -> bool = shows_ui;

on_exit_strict!(SHOWS_UI)
```
or
```rust
const SHOWS_UI: fn(&AppState) -> bool =
    |state| matches!(state, AppState::Menu | AppState::InGame { .. });

on_exit_strict!(SHOWS_UI)
```

of these, the last one is the most ergonomic ATM, and in the mean time I will change the macro to produce that, but it feels *wrong* to create a const function pointer instead of a function definition just so the traits can work...

It's worth noting - I did look into ways to make `Fn(&AppState) -> bool` work, but that kept being picked up as conflicting with `States` implementation of `StateMatcher`. I will keep looking at this in the morning